### PR TITLE
CIRCLE-15258 - Expand the definition of a Pipeline for GET Pipeline

### PR DESCRIPTION
**Pull request by gordonsyme**
Include Trigger in the GET Pipeline response.
Add a new route to get a Pipeline's configuration.

### CODE REVIEW

File: specs/v2-sketch.yaml [hunk 1]
```diff
@@ -524,3 +554,52 @@ components:
           items:
             type: string
             format: uuid
+    TriggerSummary:
+      type: object
+      properties:
+        type:
+          description: The type of the trigger
+          type: string
+          enum: ["explicit", "webhook"]
```
Comment (gordonsyme) [inline specs/v2-sketch.yaml]: These don't match the internal data, should be a 1-1 mapping between them though, and I feel consistency with the config values is important here.

File: specs/v2-sketch.yaml [hunk 2]
```diff
@@ -524,3 +554,52 @@ components:
...
+                The branch the pipeline ran on. The HEAD commit on this branch
+                was used for the pipeline, unless the `revision` parameter was
+                also set.
+              type: string
+              example: "feature/design-new-api"
+            revision:
+              description: The revision of code that the pipeline ran.
+              type: string
+              example: f454a02b5d10fcccfd7d9dd7608a76d6493a98b4
+            tag:
```
Comment (gordonsyme) [inline specs/v2-sketch.yaml]: I'm feeling less happy about the `branch`/`revision`/`tag` fields propagating everywhere now. The ways that the fields can be combined makes it hard to come up with a generic `ref` type structure that doesn't also feel forced.

File: specs/v2-sketch.yaml [hunk 3]
```diff
@@ -100,6 +100,42 @@ paths:
               operationId: getProjectById
               parameters:
                 project_slug: "$response.body#/project_slug"
+  /pipeline/{id}/config:
+    get:
+      summary: Return a pipeline's configuration
```
Comment (gordonsyme) [inline specs/v2-sketch.yaml]: I've chosen to have a single route that returns pre- and post-compile config. I don't feel strongly about this vs a separate route for each.

File: specs/v2-sketch.yaml [hunk 4]
```diff
@@ -393,11 +428,8 @@ components:
...
-        source_config:
-          type: string
-        compiled_config:
-          type: string
-        #trigger:
+        trigger:
+          $ref: "#/components/schemas/TriggerSummary"
         idempotency_key:
           type: string
         #actor:
```
Comment (gordonsyme) [inline specs/v2-sketch.yaml]: Still need to decide on the shape for an actor.

Comment (gordonsyme) [reply specs/v2-sketch.yaml]: Also, do we want to include organization data in this response?

File: specs/v2-sketch.yaml (hunk 1 above)
Comment (conormcd) [reply specs/v2-sketch.yaml]: Should we rename `explicit` to `api`? I think it's a little more self-explanatory, but I'm open to other arguments.

File: specs/v2-sketch.yaml (hunk 4 above)
Comment (conormcd) [reply specs/v2-sketch.yaml]: I would be in favour of keeping the actor as "thin" as possible here. Just a UUID for now, we can expand later.

File: specs/v2-sketch.yaml (hunk 3 above)
Comment (conormcd) [reply specs/v2-sketch.yaml]: +1 for one-for-both.

File: specs/v2-sketch.yaml (hunk 1 above)
Comment (gordonsyme) [reply specs/v2-sketch.yaml]: We have some exploratory config documentation suggesting `explicit` over `api`. I'm not wedded to either name but do feel we need to be consistent with the value used in config.

File: specs/v2-sketch.yaml (hunk 2 above)
Comment (gordonsyme) [reply specs/v2-sketch.yaml]: @ndintenfass I know you've been thinking about this before, any thoughts?

The logic is essentially:
```clojure
(or (not tag)
    (and (not branch)
         (not revision)))
```

We could add a field with a `OneOf` schema that reflects these options
* a `tag` structure
  - we build that tag
* a `branch-inclusive-or-revision` structure
  - if just the branch is specified we build HEAD of that branch
  - if just the revision is specified we build the default branch but hard reset to the revision
  - if the branch and revision are specified we build the branch hard reset to the revision
* the field is omitted
  - we build HEAD of the default branch

File: specs/v2-sketch.yaml (hunk 4 above)
Comment (gordonsyme) [reply specs/v2-sketch.yaml]: That sounds reasonable, we'll definitely want more info later.

File: specs/v2-sketch.yaml (hunk 2 above)
Comment (gordonsyme) [reply specs/v2-sketch.yaml]: I talked to Nathan about this. The executive summary is that for triggering purposes we will only support specifying the tag xor branch as top-level keys in the trigger structure.

For fetching Pipelines we will return one of the tag or branch, along with the revision that the tag/branch was resolved to.

File: specs/v2-sketch.yaml (hunk 4 above)
Comment (gordonsyme) [reply specs/v2-sketch.yaml]: Going to punt on the actor for now.

**Comment by gordonsyme** [comment]
Merging, I expect we will be iterating on it, particularly the field names and breaking out the config route into a route each for source- and compiled-config.

**Pull request merged**